notes/conds.md

```diff
@@ -0,0 +1,86 @@
+
+### Operators
+
+| Operator     | Function Signature                | Description                                | Example                                        |
+| ------------ | --------------------------------- | ------------------------------------------ | ---------------------------------------------- |
+| `/${string}` |`[] -> Any`                        | Selects value using JSON Pointer            | `["/foo/0"]`                                  |
+|              |`[] -> Any`                        | Selects whole value like "" in JSON Pointer | `[]`                                          |
+| `==`         |`[Exp<Any>, Exp<Any>] -> Bool`     | Equality                                    | `["==", ["/size"], 1]`                        |
+| `<`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than                                   | `["<",  ["/size"], 1024]`                     |
+| `<=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than or equal                          | `["<=", ["/size"], 1024]`                     |
+| `>`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than                                | `[">", ["@", "/size"], 0]`                    |
+| `>=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than or equal                       | `[">=", ["@", "/size"], 32]`                  |
+| `not`        |`[Expr<Bool>] -> Bool`             | Negation                                    | `["not", ["<=", ["/size"], 128]`              |
+| `match`      |`[Expr<Str>, Exp<Str>] -> Bool`    | String wildcard                             | `["match", ["/to"], "*@foo.com"]`             |
+| `or`         |`Exp<Bool>[] -> Bool`              | Logical **or** combinator                   | `["or", [">", ["/n"], 0], ["==", ["/n"], 0]]` |
+| `and`        |`Exp<Bool>[] -> Bool`              | Logical **and** combinator                  | `["and", [">", ["/n"], 0], ["<", ["/n"], 10]]`|
+
+## Examples
+
+Delegates `email/send` command on `did:key:alice` with a following constraint: Email is not send to address on `gmail.com`, or someone from `fission.codes` should be cc-ed.
+
+```js
+{
+  "sub": "did:key:alice",
+  "iss": "did:key:alice",
+  "aud": "did:key:bob",
+  "cmd": "email/send",
+  "cond": [
+    ["or",
+      // no recipient should be at `@gmail.com`
+      ["every", ["/to"], ["not", ["match", [], "*@gmail.com"]]],
+      // or cc should include @fission.codes
+      ["some", ["/cc"], ["match", [], "*@fission.codes"]]
+    ]
+  ]
+}
+```
+-->
+
+## logic programming language
+
+We are also considering logic programming alternative inspired by datomic query styntax.
+
+### concrete proposal for the inital release
+
+Here's a sketch that shows the features
+
+```js
+{
+  "iss": "did:key:alice",
+  // ...
+  "policy": [
+     ["some", "?x"],
+     ["$args", "foo..bar.[].baz", "?x"],
+     ["match", "?x", "*@example.com"],
+      
+     ["$args", "foo.quux", "?y"],
+     ["?y", ">", 42],
+     ["?x", "<", "?y"]
+  ]
+}
```

Comment:
Just trying to fork the conversation into a separate thread on the language. Here's my latest updated in Rust

```rust
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Expression {
    Variable(String),

    Literal(Ipld), // FIXME only JSON? Because we need a way to inspect CIDs + Bytes
    Path(Vec<Index>),

    Equal(Box<Expression>, Box<Expression>),
    Not(Box<Expression>),

    And(Vec<Expression>),
    Or(Vec<Expression>),

    GreaterThan(Box<Expression>, Box<Expression>),
    GreaterOrEqual(Box<Expression>, Box<Expression>),

    LessThan(Box<Expression>, Box<Expression>),
    LessOrEqual(Box<Expression>, Box<Expression>),

    Every(Box<Expression>),
    Some(Box<Expression>),
    // None(Box<Expression>)

    Match(String, Box<Expression>)
    Key(Variable)
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Index {
    All,
    Index(usize),
    Key(String),
    RecDesend,
}
```

We don't need a `None` relation, right? It's just `[not [some ...]]`, ya?

Reply:
> Every(Box<Expression>),
>  Some(Box<Expression>),

I am very skeptical of these, expression may refer to multiple bindings and it is not clear which one some/every would apply to

Reply:
> expression may refer to multiple bindings and it is not clear which one some/every would apply to

They don't contain expressions. They contain variables.

Reply:
Why does every contain two variables?

Reply:
Great question. I'm not thrilled with having to expose that to users, but `every` basically says "filtering is the same size as not filtering".

```js
["$.foo.bar", "?x"] // X
["$.foo.bar", "?y"] // Y
["match", "?x", "*@example.com"] // Filter ?x
["==", "?x", "?y"] // Do they still have the same size?
```

Which we could hide behind an operator

```js
["every", "?x", /* does */ "?y"]
["$.foo.bar", "?x"]
["match", "?y", "*@example.com"]
```

Reply:
Some is simpler because it just says "is not empty". It actually may be more intuitive to just let people express `some = not empty`

Reply:
I'll keep sharing my Rust schema as things evolve

```rust
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Expression {
    Args,
    Goal(Goal),
    Literal(Json), // FIXME only JSON? Because we need a way to inspect CIDs + Bytes
    Path(Vec<Index>),

    Equal(Box<Expression>, Box<Expression>),
    Not(Box<Expression>),

    And(Vec<Expression>),
    Or(Vec<Expression>),

    GreaterThan(Box<Expression>, Box<Expression>),
    GreaterOrEqual(Box<Expression>, Box<Expression>),

    LessThan(Box<Expression>, Box<Expression>),
    LessOrEqual(Box<Expression>, Box<Expression>),

    // Future
    Every(Goal, Goal), // Forall goal  in domain
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Goal(String);

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Index {
    All,
    Index(usize),
    Key(String),
    RecDesend,
}

pub enum Json {
    Null,
    Bool(bool),
    Float(f64),
    Integer(i128), // Only up to 2^53 , not i128 like in libipld? 🤔
    String(String),
    Array(Vec<Json>),
    Object(BTreeMap<String, Json>),
}
```

I think the most controversial thing here is that we don't match on IPLD, but directly on JSON (from an earlier conversation that we had). This means no `Link`, or `Bytes`. (Technically there's no integers either, but we infer them per IPLD). On the plus side, we don't have to introduce new syntax for these — you can match on bytes by looking at `./.bytes` and `match`ing on the base64.

Reply:
> Literal(Json)

Can we restrict literals to just scalars ? I think it will get hairy otherwise especially if we start extending syntax in the future.

> Select(Vec<Index>)

After some implementation experiments I came out with an opinion that I would like to have more clear way to signal whether selection should return one result or many. If you have `[]` or some form of range like `[1:9]` anywhere in the path it results in many otherwise one or none.

I think it would be less ambiguous to have JSON pointer selectors and then clause like `['in', '?item', '?list']` to reference (some) children.

It isn't a strongly held opinion, but so far I find it explicit `in` to be more comprehendible. Furthermore I don't expect to write those paths manually but rather generate them from some DSL which likely will generate will look more like to compile to less ambiguous syntax (with `in`) than use fancy jq selector.

> This means no Link, or Bytes. (Technically there's no integers either, but we infer them per IPLD). On the plus side, we don't have to introduce new syntax for these — you can match on bytes by looking at ./.bytes and matching on the base64.

I think I we need to support bytes natively. We do not use JSON in our UCANs, but rather use a DAG-CBOR which has native support for bytes. I can imagine expressions around byte prefixes / slices and having to deal with base64 string would be awkward.

I think in case of JSON bytes could simply be represented by arrays of bytes. Although in literals I would support idea of using `{ '/': { 'bytes': base64 } }`.

I have less strong opinions about links. Still I would prefer if representation was base encoding agnostic that is to say `./` should probably be bytes and not string or perhaps there is a better way but I really would dislike if base encoding has leaked somehow. Put it differently I don't think string operators should apply to links.

Reply:
I want to try separate threads on separate issues because number discussions here is becoming impossible to follow. Forked this into https://github.com/ucan-wg/delegation/issues/4 we can close it when we align